sherpa_contrib/diag_resp.py

```diff
@@ -0,0 +1,712 @@
+
+        if instrument.upper().startswith("UVOT"):
+            instrument = "UVOT"
+
+        instrument = instrument.upper()
+        detnam  = None
+
+
+    ### setup 'instfilter' argument; only Swift/UVOT energy-grid is dependent on this quantity ###
+    if telescope.lower() == "swift" and instrument.upper() == "UVOT":
+        if instfilter.upper() not in ["B","V","U","UVM2","UVW1","UVW2","WHITE"]:
+            raise ValueError("'telescope=Swift' with 'instrument=UVOT' requires the 'instfilter' argument to be B|V|U|UVM2|UVW1|UVW2|White")
+
+        instfilter = instfilter.upper()
+
+    else:
+        instfilter = None
+
+
+    return telescope, instrument, detnam, instfilter
+
+
+
+@_reformat_wmsg
+def build_resp(emin, emax, offset: int, ethresh: float|None=1e-12):
+    """
+    Return a diagonal RMF and flat ARF data objects with matching energy grid.
+    Use set_rmf and set_arf on the respective instances.
+
+    Parameters:
+        emin - array of energy grid lower bin edge
+        emax - array of energy grid upper bin edge
+        offset - integer-value starting channel number, typically 0 or 1.
+        ethresh - number or None, controls whether zero-energy bins are replaced.
+
+    """
+
+    logger = getLogger(__name__)
+
+    try:
+        flatarf = shpmkarf(emin, emax, ethresh=ethresh)
+        diag_rmf = shpmkrmf(emin, emax, startchan=offset, ethresh=ethresh)
+
+        #################################################################
+        ### remove once 'startchan' is factored into channel enumeration
+        ### in sherpa/astro/instrument.py
+
+        if offset != 1:
+            diag_rmf.f_chan += offset - 1
+
+        #################################################################
+
+    except Exception as exc:
+        if repr(exc).find("value <= 0") != -1:
+
+            raise RuntimeError(f"{exc.args[0]}.  Set 'ethresh' to a float value greater than zero and not None.") from None
+
+        logger.warning(exc)
+        raise RuntimeError(exc) from exc
+
+
+
+    wmsg = "RMF and ARF data objects returned; use 'set_rmf' and 'set_arf' to set the respective instances to dataset ID."
+
+    logger.warning("%s", f"\n{wmsg:>{len(wmsg)+4}}")
+
+
+    return diag_rmf, flatarf
+
+
+
+def mkdiagresp(telescope: str = "Chandra",
+               instrument: str = "ACIS",
+               detector: str|None = None,
+               instfilter: str|None = None,
+               refspec: str|DataPHA|None = None,
+               chantype: str = "PI",
+               nchan: int|None = None,
+               ethresh: float|None = 1e-12):
+    """
+    Return a diagonal RMF and flat ARF data objects with matching energy grid for a specified instrument/detector.
+    Use set_rmf and set_arf on the respective instances.  Use 'build_resp' for a non-instrument specific [or generalized] energy-grid configuration.
+
+    "PI" channel-type is the default, assumed spectral type.
+
+
+    Supported arguments:
+
+        telescope="Chandra"
```

Comment:
Gratings?

Reply:
I need to think this through... what would be the use case (since it's still a bit unclear to me if there's a use with the stowed background alongside dispersed data set?  That said, ObsIDs 6647 and 6648 are ACIS-S/HETG observations that had the zeroth-order coordinates incorrectly input  by the PI, so much of the field is source free...)

Also, how many diffraction orders should we go out?  It's easy enough to get 0-3; but any higher?

Reply:
IDK;  I just noticed they weren't covered.

Presumably same use case as say the other non-chandra datasets :)

In general the grating RMFs are "nearly" diagonal -- but take a while to compute; so one may want a diagonal response to get a quick-n-dirty answer.

Reply:
fair 'nuff! I think I've just settled on @DougBurke's school of thought: if you're using gratings you have too many counts to play with and it's your problem.

::continues gnawing::